### DISCUSSION TITLE
do_orm_execute and union execution state

### DISCUSSION THREAD

**Discussion by ay-ramp** [sqlalchemy/sqlalchemy#9633] [Usage Questions]
Hey there,

We're currently using SqlAlchemy 1.4.46 and seeing some unexpected behavior when combining `do_orm_execute` and 2.0 style `union` queries.

Here's an example:

```
from typing import Any

from sqlalchemy import Column, Integer, create_engine, select, event
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import Session, with_loader_criteria

Base: Any = declarative_base()

class A(Base):
    __tablename__ = "a"
    id = Column(Integer, primary_key=True)


class B(Base):
    __tablename__ = "b"
    id = Column(Integer, primary_key=True)


@event.listens_for(Session, "do_orm_execute")
def _add_soft_delete_filtering_criteria(execute_state):
    print("execute state", execute_state)
    print("has is_column_load", hasattr(execute_state, "is_column_load"))
    # if (
    #     not execute_state.is_column_load
    #     and not execute_state.is_relationship_load
    #     and not execute_state.execution_options.get("include_soft_deleted", False)
    # ):
    #     execute_state.statement = execute_state.statement.options(
    #         with_loader_criteria(
    #             ...,  # My SoftDelete mixin
    #             lambda cls: cls.deleted_at.is_(None),
    #             include_aliases=True,
    #         )
    #     )

e = create_engine("sqlite://", echo=False)
Base.metadata.create_all(e)

session = Session(e)

all_items_stmt = select(A.id).union(select(B.id))
session.execute(all_items_stmt).scalars().all()
```

When running, I only get a single output:

```
execute state <sqlalchemy.orm.session.ORMExecuteState object at 0x10293a030>
has is_column_load False
```

Which causes our original code (commented out above) to fail with an attribute error when running.

How would you advise us ensure that our global filters work with `union` queries like this? Is this an issue with our SQLAlchemy version or a usage problem on our end?

Thank you very much in advance!

**Comment by zzzeek** [comment]
two things:

1. that's a bug in ORMExecuteState (#9634), the attribute is raising an attributeerror internally which is propagated by `@property`
2. the actual query you're doing isn't loading ORM data so that's why the error is happening.

use this instead 

    print("is_column_load", execute_state.is_orm_statement and execute_state.is_column_load)

**Reply by ay-ramp** [reply]
Got it - thank you a lot! Follow-up question--

How would we apply the loader criteria (i.e. let's say A and B inherited from a `SoftDeleteMixin` and we wanted to apply the filter automatically to them) automatically for these types of queries? Would we need to do something with `before_execute`?

**Reply by zzzeek** [reply]
UNIONs are special and this is possibly a bug.

here is the test

```py
from typing import Any

from sqlalchemy import Column, Integer, create_engine, select, event, DateTime
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import Session, with_loader_criteria

Base: Any = declarative_base()

class SoftDeleted:
    deleted_at = Column(DateTime)

class A(SoftDeleted, Base):
    __tablename__ = "a"
    id = Column(Integer, primary_key=True)

class B(SoftDeleted, Base):
    __tablename__ = "b"
    id = Column(Integer, primary_key=True)


@event.listens_for(Session, "do_orm_execute")
def _add_soft_delete_filtering_criteria(execute_state):
    execute_state.statement = execute_state.statement.options(
        with_loader_criteria(
            SoftDeleted,
            lambda cls: cls.deleted_at.is_(None),
            include_aliases=True,
        )
    )

e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

session = Session(e)

all_items_stmt = select(A.id).union(select(B.id))
session.execute(all_items_stmt).scalars().all()
```

it's not applying the criteria to each sub-select.   It probably should.   

usually a UNION should be packaged into from_statement() and then an outer SELECT would select an ORM entity.  But you are selecting two different ORM entities here.

You can trick it like this for now:

    all_items_stmt = select(A.id).from_statement(select(A.id).union(select(B.id)))

but that's not a great solution.    I dont really have a fix for this right now unfortunately.

**Reply by zzzeek** [reply]
#9635

**Reply by zzzeek** [reply]
I have a good fix for #9635 coming up for 2.0.10.

**Reply by ay-ramp** [reply]
amazing - much appreciated. assuming we likely won't get a 1.4x backport?

**Reply by zzzeek** [reply]
yeah we are not backporting issues which 1. fail every time and 2. only appear with certain coding patterns.    people writing new code we would want to get onto 2.0 first before they write new things.

**Reply by ay-ramp** [reply]
totally fair. thanks as always!
